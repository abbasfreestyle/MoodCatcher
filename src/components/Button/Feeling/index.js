import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import { Animated, View, Text, StyleSheet } from 'react-native';

import { timing } from 'utils/animation';

import Touchable from 'components/Touchable';
import Grid from 'components/Grid';

import theme from 'theme';

const styles = StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'center',
    padding: 10
  },
  text: {
    color: theme.onSecondary,
    fontSize: 18,
    fontWeight: '400',
    textTransform: 'uppercase'
  }
});

class Feeling extends PureComponent {
  constructor(props) {
    super(props);
    this.color = new Animated.Value(props.selected ? 0 : 100);
  }

  componentDidUpdate(prevProps) {
    const { selected } = this.props;
    if (prevProps.selected !== selected) {
      timing(this.color, selected ? 0 : 100, 200).start();
    }
  }

  render() {
    const { children, onPress, flex, margin, ...touchableProps } = this.props;
    const backgroundColor = this.color.interpolate({
      inputRange: [0, 100],
      outputRange: [theme.secondary, theme.disabled]
    });
    return (
      <Grid.Row margin={margin}>
        <View flex={flex ? 1 : 0}>
          <Touchable onPress={onPress} borderRadius={5} {...touchableProps}>
            <Animated.View
              style={[styles.container, { backgroundColor }]}
              pointerEvents="box-only"
            >
              <Text style={styles.text}>{children}</Text>
            </Animated.View>
          </Touchable>
        </View>
      </Grid.Row>
    );
  }
}

Feeling.defaultProps = {
  flex: false,
  margin: 0,
  selected: false
};

Feeling.propTypes = {
  children: PropTypes.string.isRequired,
  onPress: PropTypes.func.isRequired,
  selected: PropTypes.bool,
  flex: PropTypes.bool,
  margin: PropTypes.number
};

export default Feeling;
