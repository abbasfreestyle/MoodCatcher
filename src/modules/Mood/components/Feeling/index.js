import React from 'react';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';

import { selectFeeling } from 'modules/Mood/selectors';
import { addFeeling, removeFeeling } from 'modules/Mood/actions';

import { Button } from 'components';

export const Feeling = props => {
  const { feeling, selected, onRemoveFeeling, onAddFeeling } = props;

  return (
    <Button.Feeling
      flex
      margin={10}
      selected={selected}
      onPress={() => {
        selected ? onRemoveFeeling(feeling.id) : onAddFeeling(feeling);
      }}
    >
      {feeling.name}
    </Button.Feeling>
  );
};

Feeling.propTypes = {
  feeling: PropTypes.object.isRequired,
  selected: PropTypes.bool.isRequired,
  onAddFeeling: PropTypes.func.isRequired,
  onRemoveFeeling: PropTypes.func.isRequired
};

const mapStateToProps = (state, props) => ({
  selected: selectFeeling(state, props.feeling.id)
});

const mapDispatchToProps = dispatch => ({
  onAddFeeling: feeling => dispatch(addFeeling(feeling)),
  onRemoveFeeling: id => dispatch(removeFeeling(id))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Feeling);
