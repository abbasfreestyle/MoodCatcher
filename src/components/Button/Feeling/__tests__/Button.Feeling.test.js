import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Feeling from '..';

const requiredProps = {
  children: 'Happy',
  onPress: jest.fn()
};

jest.useFakeTimers();

const wrapper = shallow(<Feeling {...requiredProps} />);

describe('Button > Feeling.', () => {
  describe('Snapshots.', () => {
    afterEach(() => {
      wrapper.setProps({ ...Feeling.defaultProps, ...requiredProps });
      jest.advanceTimersByTime(0);
    });

    test('should match default props', () => {
      expect(Feeling.defaultProps).toMatchSnapshot();
    });

    test('should match with default props provided', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('should match when selected is true', () => {
      wrapper.setProps({ selected: true });
      jest.advanceTimersByTime(200);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('should match when flex is true', () => {
      wrapper.setProps({ flex: true });
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('should match when margin is different', () => {
      wrapper.setProps({ margin: 20 });
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});
