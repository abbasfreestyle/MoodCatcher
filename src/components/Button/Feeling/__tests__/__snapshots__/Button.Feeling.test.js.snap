// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button > Feeling. Snapshots. should match default props 1`] = `
Object {
  "flex": false,
  "margin": 0,
  "selected": false,
}
`;

exports[`Button > Feeling. Snapshots. should match when flex is true 1`] = `
<Row
  margin={0}
>
  <View
    flex={1}
  >
    <Touchable
      borderRadius={5}
      onPress={[MockFunction]}
      rippleColor="white"
      selected={false}
    >
      <AnimatedComponent
        pointerEvents="box-only"
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "padding": 10,
            },
            Object {
              "backgroundColor": "rgba(212, 212, 213, 1)",
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 18,
              "fontWeight": "400",
              "textTransform": "uppercase",
            }
          }
        >
          Happy
        </Text>
      </AnimatedComponent>
    </Touchable>
  </View>
</Row>
`;

exports[`Button > Feeling. Snapshots. should match when margin is different 1`] = `
<Row
  margin={20}
>
  <View
    flex={0}
  >
    <Touchable
      borderRadius={5}
      onPress={[MockFunction]}
      rippleColor="white"
      selected={false}
    >
      <AnimatedComponent
        pointerEvents="box-only"
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "padding": 10,
            },
            Object {
              "backgroundColor": "rgba(212, 212, 213, 1)",
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 18,
              "fontWeight": "400",
              "textTransform": "uppercase",
            }
          }
        >
          Happy
        </Text>
      </AnimatedComponent>
    </Touchable>
  </View>
</Row>
`;

exports[`Button > Feeling. Snapshots. should match when selected is true 1`] = `
<Row
  margin={0}
>
  <View
    flex={0}
  >
    <Touchable
      borderRadius={5}
      onPress={[MockFunction]}
      rippleColor="white"
      selected={true}
    >
      <AnimatedComponent
        pointerEvents="box-only"
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "padding": 10,
            },
            Object {
              "backgroundColor": "rgba(242, 106, 163, 1)",
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 18,
              "fontWeight": "400",
              "textTransform": "uppercase",
            }
          }
        >
          Happy
        </Text>
      </AnimatedComponent>
    </Touchable>
  </View>
</Row>
`;

exports[`Button > Feeling. Snapshots. should match with default props provided 1`] = `
<Row
  margin={0}
>
  <View
    flex={0}
  >
    <Touchable
      borderRadius={5}
      onPress={[MockFunction]}
      rippleColor="white"
      selected={false}
    >
      <AnimatedComponent
        pointerEvents="box-only"
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "padding": 10,
            },
            Object {
              "backgroundColor": "rgba(212, 212, 213, 1)",
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 18,
              "fontWeight": "400",
              "textTransform": "uppercase",
            }
          }
        >
          Happy
        </Text>
      </AnimatedComponent>
    </Touchable>
  </View>
</Row>
`;
